thinc/backends/_custom_kernels.py

```diff
@@ -62,74 +62,106 @@
 )
 
 
-def _get_kernel(name):
-    """A small wrapper around KERNELS.get_function that verifies first that
-    compiler kernels are available (cupy is installed)."""
-    if KERNELS is None:
-        return None
-    else:
-        return KERNELS.get_function(name)
+class LazyKernel:
+    """Wraps around `cupy.RawModule` and `cupy.RawKernel` to verify CuPy availability
+    and lazily compile the latter on first invocation.
+
+    The default CuPy behaviour triggers the compilation as soon as the `cupy.RawKernel` object
+    is accessed."""
+
+    name: str
+    _kernel: Optional["cupy.RawKernel"]
+    _compile_callback: Optional[Callable[[], "cupy.RawKernel"]]
+    _compiled: bool
+
+    __slots__ = ["name", "_kernel", "_compile_callback", "_compiled"]
+
+    def __init__(
+        self,
+        name: str,
+        *,
+        compile_callback: Optional[Callable[[], "cupy.RawKernel"]] = None,
+    ) -> None:
+        self.name = name
+        self._kernel = None
+        self._compile_callback = compile_callback
+        self._compiled = False
+
+    def __call__(self, *args, **kwargs):
+        self._compile_kernel()
+        if self._kernel is None:
+            raise ValueError(
+                f"couldn't invoke Cupy kernel '{self.name}' as it was not compiled"
+            )
+        self._kernel(*args, **kwargs)
+
+    def _compile_kernel(self):
+        if self._compiled:
+            return
+
+        if self._compile_callback is not None:
+            self._kernel = self._compile_callback()
+        elif KERNELS is not None:
+            self._kernel = KERNELS.get_function(self.name)
+        self._compiled = True
```

Comment:
We always have `KERNELS` I think? So both code paths should return something (or raise an exception). I am wondering if we can't do without `_compiled` and signify that `self._kernel is None` means that the kernel is not yet compiled. But I might definitely be overlooking something here.

Reply:
Looking at the initialization code, it seems like `KERNELS` is `None` when `has_cupy_gpu` is `False`. But it is true that `CupyOps` won't ever be called in that case, so I think we can get rid of the flag.